README.md
# flutter_trading_chart

Flutter package to display fully featured trading charts.  
Can be used to display candlestick charts and add line series, scatter series, volume bars but also the volume profile.

![example](https://github.com/AmauryBuguet/flutter_trading_chart/blob/master/example/Pictures/example.png)
## Getting Started
### Install
```yaml
dependencies:
  flutter_trading_chart: ^0.0.1
```

or use
```
flutter pub add flutter_trading_chart
```

### Usage
```dart
import 'dart:math';

import 'package:collection/collection.dart';
import 'package:flutter/material.dart';
import 'package:flutter_trading_chart/flutter_trading_chart.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: ExempleTradingChart(),
    );
  }
}

class ExempleTradingChart extends StatefulWidget {
  const ExempleTradingChart({Key? key}) : super(key: key);

  @override
  _ExempleTradingChartState createState() => _ExempleTradingChartState();
}

class _ExempleTradingChartState extends State<ExempleTradingChart> {
  TradingChartController controller = TradingChartController(
    settings: TradingChartSettings(
      volume: VolumeSettings(
        visible: true,
      ),
      vpvr: VpvrSettings(
        visible: true,
      ),
    ),
  );
  @override
  void initState() {
    super.initState();
    setCandleSerie();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("TRADING CHART EXEMPLE"),
      ),
      body: TradingChart(
        controller: controller,
      ),
    );
  }

  CandlestickSerie generateCandleStickSerie() {
    var rng = Random();
    double open = 50000;
    int timestamp = 1621184400000;
    double volume = 1000;
    List<Candlestick> candles = [];
    for (int i = 0; i < 100; ++i) {
      double close = open * (1 + (rng.nextInt(11) - 4.5) / 100);
      double high = max(close, open) * (1 + rng.nextInt(2) / 100);
      double low = min(close, open) * (1 - rng.nextInt(2) / 100);
      volume = volume * (1 + (rng.nextInt(11) - 4.8) / 100);
      candles.add(Candlestick(
        timestamp: timestamp + i * 300000,
        open: open,
        high: high,
        low: low,
        close: close,
        volume: volume,
      ));
      open = close;
    }
    return CandlestickSerie(
      name: "Price",
      candles: candles,
    );
  }

  void setCandleSerie() {
    controller.data.candleSerie = generateCandleStickSerie();
    List<Point> movingAveragePoints = [];
    for (int i = 8; i < controller.data.candleSerie!.candles.length; ++i) {
      final sublist = controller.data.candleSerie!.candles.sublist(i - 8, i + 1).map((e) => e.close).toList();
      movingAveragePoints.add(Point(
        timestamp: controller.data.candleSerie!.candles[i].timestamp,
        y: sublist.average,
      ));
    }
    controller.data.lineSeries.add(LineSerie(
      name: "Moving Average",
      color: Colors.yellow,
      points: movingAveragePoints,
    ));
    controller.setTimestamps(
      controller.data.candleSerie!.candles.first.timestamp,
      controller.data.candleSerie!.candles.last.timestamp,
    );
  }
}
```



CHANGELOG.md
## 0.0.2

* Fixed example and improved description

## 0.0.1

* A Flutter package to display fully featured trading charts.


lib/src/series.dart
import 'package:flutter/material.dart';

class Candlestick {
  final int timestamp;
  final double open;
  double high;
  double low;
  double close;
  double volume;

  Candlestick({
    required this.timestamp,
    required this.open,
    required this.high,
    required this.low,
    required this.close,
    required this.volume,
  });
}

class Point {
  final int timestamp;
  double y;

  Point({required this.timestamp, required this.y});
}

class CandlestickSerie {
  final String name;
  List<Candlestick> candles;
  double wickWidth;
  double ratioCandleSpace;
  Paint bullPaint;
  Paint bearPaint;
  Paint dojiPaint;

  CandlestickSerie({
    required this.name,
    List<Candlestick>? candles,
    this.wickWidth = 1.0,
    this.ratioCandleSpace = 0.8,
    bullPaint,
    bearPaint,
    dojiPaint,
  })  : bullPaint = bullPaint ?? Paint()
          ..color = Colors.white
          ..style = PaintingStyle.stroke,
        bearPaint = bearPaint ?? Paint()
          ..color = Colors.red
          ..style = PaintingStyle.stroke,
        dojiPaint = dojiPaint ?? Paint()
          ..color = Colors.blue
          ..style = PaintingStyle.stroke,
        candles = candles ?? [];
}

class LineSerie {
  final String name;
  final Color color;
  List<Point> points;

  LineSerie({
    required this.name,
    required this.color,
    List<Point>? points,
  }) : points = points ?? [];
}

class ScatterSerie {
  final String name;
  final Color color;
  final double pointSize;
  List<Point> points;

  ScatterSerie({
    required this.name,
    List<Point>? points,
    required this.color,
    this.pointSize = 10,
  }) : points = points ?? [];
}

class StraightLineSerie {
  final String name;
  double y;
  final Color color;
  bool dotted;
  double width;

  StraightLineSerie({
    required this.name,
    required this.y,
    required this.color,
    this.dotted = false,
    this.width = 1.0,
  });
}

class BarSerie {
  final String name;
  final Color color;
  List<Point> values;

  BarSerie({
    required this.name,
    required this.color,
    List<Point>? values,
  }) : values = values ?? [];
}


lib/src/trading_chart.dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

import 'trading_chart_painter.dart';
import 'trading_chart_utils.dart';

class Tradding extends StatefulWidget {
  const Tradding({Key? key}) : super(key: key);

  @override
  _TraddingState createState() => _TraddingState();
}

class _TraddingState extends State<Tradding> {
  @override
  Widget build(BuildContext context) {
    return Container();
  }
}

class TradingChart extends StatefulWidget {
  final TradingChartController controller;
  const TradingChart({Key? key, required this.controller}) : super(key: key);

  @override
  _TradingChartState createState() => _TradingChartState();
}

class _TradingChartState extends State<TradingChart> {
  final Map<int, Offset> _ptrsMap = {};
  @override
  Widget build(BuildContext context) {
    return Container(
      color: widget.controller.settings.backgroundColor,
      child: Listener(
        onPointerDown: (event) {
          _ptrsMap[event.pointer] = event.localPosition;
        },
        onPointerMove: (event) {
          if (_ptrsMap.length == 1) {
            widget.controller.endTsNotifier.value -=
                event.delta.dx ~/ widget.controller.pixelsPerMs;
            widget.controller.startTsNotifier.value -=
                event.delta.dx ~/ widget.controller.pixelsPerMs;
          } else if (_ptrsMap.length == 2) {
            Offset otherPtr = _ptrsMap.entries
                .firstWhere((element) => element.key != event.pointer)
                .value;
            Offset oldPtr = _ptrsMap.entries
                .firstWhere((element) => element.key == event.pointer)
                .value;
            double distance = (event.localPosition.dx - otherPtr.dx).abs() -
                (oldPtr.dx - otherPtr.dx).abs();
            double coef =
                (otherPtr.dx + event.localPosition.dx) / (context.size!.width);
            widget.controller.endTsNotifier.value -=
                (2 - coef) * distance ~/ widget.controller.pixelsPerMs;
            widget.controller.startTsNotifier.value +=
                coef * distance ~/ widget.controller.pixelsPerMs;
            _ptrsMap[event.pointer] = event.localPosition;
          }
        },
        onPointerUp: (event) {
          _ptrsMap.remove(event.pointer);
        },
        onPointerSignal: (event) {
          if (event is PointerScrollEvent) {
            double coef = event.localPosition.dx * 2 / (context.size!.width);
            widget.controller.endTsNotifier.value += 4 *
                (2 - coef) *
                event.scrollDelta.dy ~/
                widget.controller.pixelsPerMs;
            widget.controller.startTsNotifier.value -= 4 *
                coef *
                event.scrollDelta.dy ~/
                widget.controller.pixelsPerMs;
          }
        },
        child: CustomPaint(
          size: Size.infinite,
          painter: TradingChartPainter(
            endTimestamp: widget.controller.endTsNotifier,
            startTimestamp: widget.controller.startTsNotifier,
            controller: widget.controller,
          ),
        ),
      ),
    );
  }
}


lib/src/trading_chart_utils.dart
import 'package:flutter/material.dart';

import 'series.dart';

class TradingChartData {
  CandlestickSerie? candleSerie;
  List<LineSerie> lineSeries;
  List<ScatterSerie> scatterSeries;
  List<StraightLineSerie> straightLineSeries;

  TradingChartData({
    this.candleSerie,
    List<LineSerie>? lineSeries,
    List<ScatterSerie>? scatterSeries,
    List<StraightLineSerie>? straightLineSeries,
  })  : lineSeries = lineSeries ?? [],
        scatterSeries = scatterSeries ?? [],
        straightLineSeries = straightLineSeries ?? [];
}

class LRTB {
  double left = 0.1;
  double right = 0.1;
  double top = 0.1;
  double bottom = 0.1;

  LRTB({
    this.left = 0.1,
    this.right = 0.1,
    this.bottom = 0.1,
    this.top = 0.1,
  });
}

class VpvrSettings {
  bool visible;
  int nbBars;
  double maxPercentWidth;
  Color color;
  Color highestColor;

  VpvrSettings({
    this.visible = false,
    this.nbBars = 50,
    this.maxPercentWidth = 0.3,
    this.color = Colors.blue,
    this.highestColor = Colors.white,
  });
}

class VolumeSettings {
  bool visible;
  double maxPercentHeight;
  Color upColor;
  Color downColor;

  VolumeSettings({
    this.visible = false,
    this.maxPercentHeight = 0.1,
    this.upColor = Colors.white,
    this.downColor = Colors.red,
  });
}

class TradingChartSettings {
  bool gridVisible;
  VpvrSettings vpvr;
  VolumeSettings volume;
  double pricePercentMargin;
  LRTB chartMargins;
  Color backgroundColor;
  int nbCandlesInitiallyDisplayed;
  int maxCandlesDisplayed;
  int nbDivisionsXAxis;
  int nbDivisionsYAxis;
  Paint linesPaint;
  Paint gridPaint;

  TradingChartSettings({
    this.gridVisible = true,
    VpvrSettings? vpvr,
    VolumeSettings? volume,
    this.pricePercentMargin = 10.0,
    chartMargins,
    this.backgroundColor = const Color.fromARGB(255, 35, 35, 55),
    this.nbCandlesInitiallyDisplayed = 100,
    this.maxCandlesDisplayed = 300,
    this.nbDivisionsXAxis = 5,
    this.nbDivisionsYAxis = 5,
    linesPaint,
    gridPaint,
  })  : chartMargins = chartMargins ?? LRTB(),
        linesPaint = linesPaint ?? Paint()
          ..color = Colors.white
          ..style = PaintingStyle.stroke,
        gridPaint = gridPaint ?? Paint()
          ..color = const Color.fromARGB(25, 255, 255, 255)
          ..style = PaintingStyle.stroke,
        vpvr = vpvr ?? VpvrSettings(),
        volume = volume ?? VolumeSettings();
}

class TradingChartController {
  TradingChartData data;
  TradingChartSettings settings;
  ValueNotifier<int> startTsNotifier;
  ValueNotifier<int> endTsNotifier;
  double pixelsPerMs = 0;
  double pixelsPerUSDT = 0;
  double maxY = 0;
  double minY = 0;
  int displayedCandlesOnReset = 50;
  Size size = const Size(1, 1);

  TradingChartController({
    TradingChartData? data,
    TradingChartSettings? settings,
    int? startTs,
    int? endTs,
  })  : startTsNotifier = ValueNotifier(startTs ?? 0),
        endTsNotifier = ValueNotifier(endTs ?? 1),
        data = data ?? TradingChartData(),
        settings = settings ?? TradingChartSettings();

  Offset chartPositionToLocalPosition(Point pt) {
    return Offset(
      (pt.timestamp - startTsNotifier.value) * pixelsPerMs +
          settings.chartMargins.left * size.width,
      (maxY - pt.y) * pixelsPerUSDT + settings.chartMargins.top * size.height,
    );
  }

  Point localPositionToChartPosition(Offset pt) {
    return Point(
      timestamp:
          (pt.dx - settings.chartMargins.left * size.width) ~/ pixelsPerMs +
              startTsNotifier.value,
      y: maxY -
          (pt.dy - settings.chartMargins.top * size.height) / pixelsPerUSDT,
    );
  }

  void setTimestamps(int start, int end) {
    startTsNotifier.value = start;
    endTsNotifier.value = end;
  }

  void centerChart() {
    if (data.candleSerie != null && data.candleSerie!.candles.isNotEmpty) {
      if (data.candleSerie!.candles.length > displayedCandlesOnReset) {
        startTsNotifier.value = data
            .candleSerie!
            .candles[data.candleSerie!.candles.length - displayedCandlesOnReset]
            .timestamp;
      } else {
        startTsNotifier.value = data.candleSerie!.candles.first.timestamp;
      }
      endTsNotifier.value = data.candleSerie!.candles.last.timestamp;
    }
  }
}


lib/src/trading_chart_painter.dart
import 'dart:math';
import 'dart:ui';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:path_drawing/path_drawing.dart';

import 'series.dart';
import 'trading_chart_utils.dart';

class TradingChartPainter extends CustomPainter {
  final TradingChartController controller;
  ValueNotifier<int> startTimestamp;
  ValueNotifier<int> endTimestamp;

  TradingChartPainter({
    required this.controller,
    required this.startTimestamp,
    required this.endTimestamp,
  }) : super(repaint: startTimestamp);

  @override
  void paint(Canvas canvas, Size size) {
    double marginLeft = size.width * controller.settings.chartMargins.left;
    double marginRight = size.width * controller.settings.chartMargins.right;
    double marginTop = size.height * controller.settings.chartMargins.top;
    double marginBottom = size.height * controller.settings.chartMargins.bottom;
    controller.size = size;

    // Draw grid and lines
    canvas.drawRect(
      Rect.fromLTRB(
        marginLeft,
        marginTop,
        size.width - marginRight,
        size.height - marginBottom,
      ),
      controller.settings.linesPaint,
    );

    // generate sublists and get max and min
    controller.maxY = double.negativeInfinity;
    controller.minY = double.infinity;
    CandlestickSerie? candleSerieSublist;
    var serie = controller.data.candleSerie;
    if (serie != null) {
      int firstIndex =
          serie.candles.indexWhere((e) => e.timestamp >= startTimestamp.value);
      int lastIndex = serie.candles
          .indexWhere((e) => e.timestamp >= endTimestamp.value, firstIndex);
      if (firstIndex != -1) {
        var candles = serie.candles.sublist(
            firstIndex, lastIndex != -1 ? lastIndex : serie.candles.length);
        candleSerieSublist = CandlestickSerie(
          name: serie.name,
          bearPaint: serie.bearPaint,
          bullPaint: serie.bullPaint,
          candles: candles,
          dojiPaint: serie.dojiPaint,
          ratioCandleSpace: serie.ratioCandleSpace,
          wickWidth: serie.wickWidth,
        );
        double maxi = candles.map<double>((e) => e.high).reduce(max);
        double mini = candles.map<double>((e) => e.low).reduce(min);
        if (maxi > controller.maxY) controller.maxY = maxi;
        if (mini < controller.minY) controller.minY = mini;
      }
    }
    List<LineSerie> lineSeriesSublist = [];
    for (LineSerie serie in controller.data.lineSeries) {
      if (serie.points.isNotEmpty) {
        var pts = serie.points.where((e) {
          return (e.timestamp >= startTimestamp.value) &&
              (e.timestamp <= endTimestamp.value);
        }).toList();
        if (pts.isNotEmpty) {
          lineSeriesSublist.add(LineSerie(
            name: serie.name,
            color: serie.color,
            points: pts,
          ));
          double maxi = pts.map<double>((e) => e.y).reduce(max);
          double mini = pts.map<double>((e) => e.y).reduce(min);
          if (maxi > controller.maxY) controller.maxY = maxi;
          if (mini < controller.minY) controller.minY = mini;
        }
      }
    }
    List<ScatterSerie> scatterSerieSublist = [];
    for (ScatterSerie serie in controller.data.scatterSeries) {
      if (serie.points.isNotEmpty) {
        var pts = serie.points.where((e) {
          return (e.timestamp >= startTimestamp.value) &&
              (e.timestamp <= endTimestamp.value);
        }).toList();
        if (pts.isNotEmpty) {
          scatterSerieSublist.add(ScatterSerie(
            name: serie.name,
            pointSize: serie.pointSize,
            color: serie.color,
            points: pts,
          ));
          double maxi = pts.map<double>((e) => e.y).reduce(max);
          double mini = pts.map<double>((e) => e.y).reduce(min);
          if (maxi > controller.maxY) controller.maxY = maxi;
          if (mini < controller.minY) controller.minY = mini;
        }
      }
    }
    for (var serie in controller.data.straightLineSeries) {
      if (serie.y > controller.maxY) controller.maxY = serie.y;
      if (serie.y < controller.minY) controller.minY = serie.y;
    }

    // check if there is something to display
    if (candleSerieSublist == null &&
        lineSeriesSublist.isEmpty &&
        scatterSerieSublist.isEmpty &&
        controller.data.straightLineSeries.isEmpty) {
      return;
    }

    // set utils variables
    double margin = controller.settings.pricePercentMargin *
        (controller.maxY - controller.minY) /
        100;
    controller.maxY += margin;
    controller.minY -= margin;
    controller.pixelsPerMs = (size.width - (marginLeft + marginRight)) /
        (endTimestamp.value - startTimestamp.value);
    controller.pixelsPerUSDT = (size.height - (marginBottom + marginTop)) /
        (controller.maxY - controller.minY);

    // Draw candles
    if (candleSerieSublist != null) {
      int msPerCandle =
          60000; // If there is only one candle, then it is considered as a 5-min candle.
      if (candleSerieSublist.candles.length >= 2) {
        msPerCandle = candleSerieSublist.candles[1].timestamp -
            candleSerieSublist.candles[0].timestamp;
      }
      double bodySize = ((size.width - (marginLeft + marginRight)) *
              msPerCandle *
              candleSerieSublist.ratioCandleSpace) /
          (controller.endTsNotifier.value - controller.startTsNotifier.value);
      for (Candlestick candle in candleSerieSublist.candles) {
        Paint paint;
        if (candle.open < candle.close) {
          paint = candleSerieSublist.bullPaint;
        } else if (candle.open > candle.close) {
          paint = candleSerieSublist.bearPaint;
        } else {
          paint = candleSerieSublist.dojiPaint;
        }
        // Draw up wick
        canvas.drawRect(
          Rect.fromLTRB(
            (candle.timestamp - startTimestamp.value) * controller.pixelsPerMs -
                (candleSerieSublist.wickWidth / 2) +
                marginLeft,
            (controller.maxY - candle.high) * controller.pixelsPerUSDT +
                marginTop,
            (candle.timestamp - startTimestamp.value) * controller.pixelsPerMs +
                (candleSerieSublist.wickWidth / 2) +
                marginLeft,
            (controller.maxY - max(candle.open, candle.close)) *
                    controller.pixelsPerUSDT +
                marginTop,
          ),
          paint,
        );
        // Draw down wick
        canvas.drawRect(
          Rect.fromLTRB(
            (candle.timestamp - startTimestamp.value) * controller.pixelsPerMs -
                (candleSerieSublist.wickWidth / 2) +
                marginLeft,
            (controller.maxY - min(candle.open, candle.close)) *
                    controller.pixelsPerUSDT +
                marginTop,
            (candle.timestamp - startTimestamp.value) * controller.pixelsPerMs +
                (candleSerieSublist.wickWidth / 2) +
                marginLeft,
            (controller.maxY - candle.low) * controller.pixelsPerUSDT +
                marginTop,
          ),
          paint,
        );

        // Draw body
        canvas.drawRect(
          Rect.fromLTRB(
            (candle.timestamp - startTimestamp.value) * controller.pixelsPerMs -
                (bodySize / 2) +
                marginLeft,
            (controller.maxY - candle.close) * controller.pixelsPerUSDT +
                marginTop,
            (candle.timestamp - startTimestamp.value) * controller.pixelsPerMs +
                (bodySize / 2) +
                marginLeft,
            (controller.maxY - candle.open) * controller.pixelsPerUSDT +
                marginTop,
          ),
          paint,
        );

        // draw volume
        if (controller.settings.volume.visible) {
          double maxVol =
              candleSerieSublist.candles.map((e) => e.volume).reduce(max);
          double highSize = controller.settings.volume.maxPercentHeight *
              (size.height - marginBottom - marginTop);
          canvas.drawRect(
            Rect.fromLTRB(
              (candle.timestamp - startTimestamp.value) *
                      controller.pixelsPerMs -
                  (bodySize / 2) +
                  marginLeft,
              size.height - marginBottom - (candle.volume * highSize / maxVol),
              (candle.timestamp - startTimestamp.value) *
                      controller.pixelsPerMs +
                  (bodySize / 2) +
                  marginLeft,
              size.height - marginBottom,
            ),
            Paint()
              ..color = candle.close >= candle.open
                  ? controller.settings.volume.upColor
                  : controller.settings.volume.downColor,
          );
        }
      }
    }

    // Draw Line Series
    for (LineSerie serie in lineSeriesSublist) {
      canvas.drawPoints(
          PointMode.polygon,
          serie.points
              .map((point) => Offset(
                    (point.timestamp - startTimestamp.value) *
                            controller.pixelsPerMs +
                        marginLeft,
                    (controller.maxY - point.y) * controller.pixelsPerUSDT +
                        marginTop,
                  ))
              .toList(),
          Paint()
            ..color = serie.color
            ..style = PaintingStyle.fill);
    }

    // Draw Scatter Series
    for (ScatterSerie serie in scatterSerieSublist) {
      canvas.drawPoints(
        PointMode.points,
        serie.points
            .map((e) => Offset(
                  (e.timestamp - startTimestamp.value) *
                          controller.pixelsPerMs +
                      marginLeft,
                  (controller.maxY - e.y) * controller.pixelsPerUSDT +
                      marginTop,
                ))
            .toList(),
        Paint()
          ..color = serie.color
          ..strokeCap = StrokeCap.round
          ..strokeWidth = serie.pointSize,
      );
    }

    // Draw StraightLine Series
    for (StraightLineSerie serie in controller.data.straightLineSeries) {
      var path = Path();
      path.moveTo(
        marginLeft,
        (controller.maxY - serie.y) * controller.pixelsPerUSDT + marginTop,
      );
      path.lineTo(
        size.width - marginRight,
        (controller.maxY - serie.y) * controller.pixelsPerUSDT + marginTop,
      );
      canvas.drawPath(
        serie.dotted
            ? dashPath(
                path,
                dashArray: CircularIntervalList<double>(<double>[15.0, 10.0]),
              )
            : path,
        Paint()
          ..color = serie.color
          ..style = PaintingStyle.stroke
          ..strokeWidth = serie.width,
      );
    }

    // Draw vpvr
    if (controller.settings.vpvr.visible && candleSerieSublist != null) {
      Map<int, double> volumeMap = {};
      double wideness =
          (controller.maxY - controller.minY) / controller.settings.vpvr.nbBars;
      for (Candlestick candle in candleSerieSublist.candles) {
        int level = (candle.close - controller.minY) ~/ wideness;
        volumeMap[level] = (volumeMap[level] ?? 0) + candle.volume;
      }
      double maxVol = volumeMap.values.reduce(max);
      double highSize = controller.settings.vpvr.maxPercentWidth *
          (size.width - marginLeft - marginRight);
      double barWidth = (size.height - marginBottom - marginTop) *
          0.4 /
          controller.settings.vpvr.nbBars;
      for (var entry in volumeMap.entries) {
        canvas.drawRect(
          Rect.fromLTRB(
            size.width - marginRight - (entry.value * highSize / maxVol),
            (controller.maxY - (entry.key * wideness + controller.minY)) *
                    controller.pixelsPerUSDT +
                marginTop -
                barWidth,
            size.width - marginRight,
            (controller.maxY - (entry.key * wideness + controller.minY)) *
                    controller.pixelsPerUSDT +
                marginTop +
                barWidth,
          ),
          Paint()
            ..color = entry.value == maxVol
                ? controller.settings.vpvr.highestColor
                : controller.settings.vpvr.color,
        );
      }
    }

    // hide overlappping candles
    canvas.drawRect(
        Rect.fromLTRB(
          marginLeft - 1,
          marginTop,
          0,
          size.height,
        ),
        Paint()..color = controller.settings.backgroundColor);
    canvas.drawRect(
        Rect.fromLTRB(
          size.width - marginRight + 1,
          marginTop,
          size.width,
          size.height,
        ),
        Paint()..color = controller.settings.backgroundColor);

    // Draw Y axis labels
    double intervalY = (size.height - (marginTop + marginBottom)) /
        controller.settings.nbDivisionsYAxis;
    const textStyle = TextStyle(
      color: Colors.white,
      fontSize: 10,
    );
    for (int i = 0; i < controller.settings.nbDivisionsYAxis + 1; ++i) {
      double yPos = marginTop + i * intervalY;
      canvas.drawLine(
        Offset(marginLeft - 5, yPos),
        Offset(marginLeft, yPos),
        controller.settings.linesPaint,
      );
      if (controller.settings.gridVisible) {
        canvas.drawLine(
          Offset(size.width - marginRight, yPos),
          Offset(marginTop, yPos),
          controller.settings.gridPaint,
        );
      }

      double y =
          (marginTop - yPos) / controller.pixelsPerUSDT + controller.maxY;
      final textSpan = TextSpan(
        text: y.toStringAsFixed(2),
        style: textStyle,
      );
      final textPainter = TextPainter(
        text: textSpan,
        textDirection: TextDirection.ltr,
        textAlign: TextAlign.end,
      );
      textPainter.layout(
        minWidth: 40,
        maxWidth: 50,
      );
      textPainter.paint(
        canvas,
        Offset(marginLeft - 50, yPos - 7),
      );
    }

    // Draw X axis labels
    double intervalX = (size.width - (marginLeft + marginRight)) /
        controller.settings.nbDivisionsXAxis;
    for (int i = 0; i < controller.settings.nbDivisionsXAxis + 1; ++i) {
      double xPos = marginLeft + i * intervalX;
      canvas.drawLine(
        Offset(xPos, size.height - marginBottom + 5),
        Offset(xPos, size.height - marginBottom),
        controller.settings.linesPaint,
      );
      if (controller.settings.gridVisible) {
        canvas.drawLine(
          Offset(xPos, marginTop),
          Offset(xPos, size.height - marginBottom),
          controller.settings.gridPaint,
        );
      }

      int x = (endTimestamp.value - startTimestamp.value) *
              i ~/
              controller.settings.nbDivisionsXAxis +
          startTimestamp.value;
      final textSpan = TextSpan(
        text: getStrFromDate(x),
        style: textStyle,
      );
      final textPainter = TextPainter(
        text: textSpan,
        textDirection: TextDirection.ltr,
        textAlign: TextAlign.center,
      );
      textPainter.layout(
        minWidth: 60,
        maxWidth: 60,
      );
      textPainter.paint(
        canvas,
        Offset(xPos - 30, size.height - marginBottom + 5),
      );
    }
  }

  String getStrFromDate(int ts) {
    DateTime date = DateTime.fromMillisecondsSinceEpoch(ts);
    DateTime startDate =
        DateTime.fromMillisecondsSinceEpoch(startTimestamp.value);
    DateTime endDate = DateTime.fromMillisecondsSinceEpoch(endTimestamp.value);
    if (startDate.year != endDate.year) {
      return "${date.day.toString().padLeft(2, '0')}/${date.month.toString().padLeft(2, '0')}/${date.year}";
    } else if (startDate.month != endDate.month) {
      return "${date.day.toString().padLeft(2, '0')} ${monthToStr(date.month)}";
    } else if (startDate.day != endDate.day) {
      return "${date.day.toString().padLeft(2, '0')}/${date.month.toString().padLeft(2, '0')}\n${date.hour.toString().padLeft(2, '0')}:${date.minute.toString().padLeft(2, '0')}";
    } else {
      return "${date.hour.toString().padLeft(2, '0')}:${date.minute.toString().padLeft(2, '0')}";
    }
  }

  String monthToStr(int m) {
    switch (m) {
      case 1:
        return "Jan";
      case 2:
        return "Feb";
      case 3:
        return "Mar";
      case 4:
        return "Apr";
      case 5:
        return "May";
      case 6:
        return "Jun";
      case 7:
        return "Jul";
      case 8:
        return "Aug";
      case 9:
        return "Sep";
      case 10:
        return "Oct";
      case 11:
        return "Nov";
      case 12:
        return "Dec";
      default:
        return "Unk";
    }
  }

  @override
  bool shouldRepaint(TradingChartPainter oldDelegate) {
    return true;
  }
}


example/example.dart
import 'dart:math';

import 'package:collection/collection.dart';
import 'package:flutter/material.dart';
import 'package:flutter_trading_chart/flutter_trading_chart.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: ExempleTradingChart(),
    );
  }
}

class ExempleTradingChart extends StatefulWidget {
  const ExempleTradingChart({Key? key}) : super(key: key);

  @override
  _ExempleTradingChartState createState() => _ExempleTradingChartState();
}

class _ExempleTradingChartState extends State<ExempleTradingChart> {
  TradingChartController controller = TradingChartController(
    settings: TradingChartSettings(
      volume: VolumeSettings(
        visible: true,
      ),
      vpvr: VpvrSettings(
        visible: true,
      ),
    ),
  );
  @override
  void initState() {
    super.initState();
    setCandleSerie();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("TRADING CHART EXEMPLE"),
      ),
      body: TradingChart(
        controller: controller,
      ),
    );
  }

  CandlestickSerie generateCandleStickSerie() {
    var rng = Random();
    double open = 50000;
    int timestamp = 1621184400000;
    double volume = 1000;
    List<Candlestick> candles = [];
    for (int i = 0; i < 100; ++i) {
      double close = open * (1 + (rng.nextInt(11) - 4.5) / 100);
      double high = max(close, open) * (1 + rng.nextInt(2) / 100);
      double low = min(close, open) * (1 - rng.nextInt(2) / 100);
      volume = volume * (1 + (rng.nextInt(11) - 4.8) / 100);
      candles.add(Candlestick(
        timestamp: timestamp + i * 300000,
        open: open,
        high: high,
        low: low,
        close: close,
        volume: volume,
      ));
      open = close;
    }
    return CandlestickSerie(
      name: "Price",
      candles: candles,
    );
  }

  void setCandleSerie() {
    controller.data.candleSerie = generateCandleStickSerie();
    List<Point> movingAveragePoints = [];
    for (int i = 8; i < controller.data.candleSerie!.candles.length; ++i) {
      final sublist = controller.data.candleSerie!.candles.sublist(i - 8, i + 1).map((e) => e.close).toList();
      movingAveragePoints.add(Point(
        timestamp: controller.data.candleSerie!.candles[i].timestamp,
        y: sublist.average,
      ));
    }
    controller.data.lineSeries.add(LineSerie(
      name: "Moving Average",
      color: Colors.yellow,
      points: movingAveragePoints,
    ));
    controller.setTimestamps(
      controller.data.candleSerie!.candles.first.timestamp,
      controller.data.candleSerie!.candles.last.timestamp,
    );
  }
}
